### DISCUSSION TITLE
DB_write Issue in Snap7

### DISCUSSION THREAD

**Discussion by INDRANI125** [gijzelaerr/python-snap7#291] [Q&A]
While writing String from my Python application to TIA Portal, I am not getting the exact string in my TIA portal as well as in Python program as attached. Where as it is working fine in case of read operation. Can anyone help me out?It is getting random values.
![tia_snap](https://user-images.githubusercontent.com/82458508/122713461-5b586300-d283-11eb-92f3-f31e762d1d89.PNG)

Python Code-
import snap7
IP='192.168.0.1'
RACK=0
SLOT=1
DB_NUMBER=100
START_ADDRESS=1
SIZE=259

plc=snap7.client.Client()
plc.connect(IP,RACK,SLOT)
plc_info=plc.get_cpu_info()
print(f'Module Type: {plc_info.ModuleTypeName}')
print(f'State: {plc.get_cpu_state()}')
db=plc.db_read(DB_NUMBER,START_ADDRESS,SIZE)
plc.db_write(DB_NUMBER, START_ADDRESS, b'Python to TIA')
product_name=db[0:255].decode('UTF-8').strip('\x00')
print(f'Product Name: {product_name}')
product_value=int.from_bytes(db[255:257], byteorder='big')
print(f'Product Value: {product_value}')
product_status=bool(db[257])
print(f'Product Status: {product_status}')
Python Output-
Module Type: b'CPU 1510SP-1 PN'
State: S7CpuStatusRun
Product Name: TIA Portal to Python // problem in writing this output
Product Value: 14
Product Status: True

**Comment by nikteliy** [comment]
The first two bytes in the 'String' data type contain additional information about the size of the string. You should use `get_string()` and `set_string()` functions to work with 'String' data type. I modified your code a bit:

```
import snap7
from snap7.util import set_string, get_string

IP = '192.168.0.1'
RACK = 0
SLOT = 1
DB_NUMBER = 100
START_ADDRESS = 0
SIZE = 259

plc = snap7.client.Client()
plc.connect(IP, RACK, SLOT)

plc_info = plc.get_cpu_info()
print(f'Module Type: {plc_info.ModuleTypeName}')
print(f'State: {plc.get_cpu_state()}')

string_to_db = 'Python to TIA'
# create a buffer and set buffer size equal to your string length plus 2 bytes for additional info 
buffer = bytearray(len(string_to_db) + 2)
# write string value to the buffer
set_string(buffer, 0, string_to_db, 254)
# now you are ready to write data to the plc
plc.db_write(DB_NUMBER, START_ADDRESS, buffer)

db = plc.db_read(DB_NUMBER,START_ADDRESS, SIZE)

# get the string value from the String data type
product_name = get_string(db[0:255], 0, 254)

print(f'Product Name: {product_name}')
product_value=int.from_bytes(db[255:257], byteorder='big')
print(f'Product Value: {product_value}')
product_status=bool(db[257])
print(f'Product Status: {product_status}')
```

**Reply by INDRANI125** [reply]
Hi..
Thank you for your reply.I tried your code and I am getting value in Python Output but not in TIA Portal.

**Reply by nikteliy** [reply]
Yup, I forgot to get the maximum length of the string (it's stored in the first byte of the string). Try the code below:

```
import snap7
from snap7.util import set_string, get_string

IP = '192.168.0.1'
RACK = 0
SLOT = 1
DB_NUMBER = 100
START_ADDRESS = 0
SIZE = 259

plc = snap7.client.Client()
plc.connect(IP, RACK, SLOT)

plc_info = plc.get_cpu_info()
print(f'Module Type: {plc_info.ModuleTypeName}')
print(f'State: {plc.get_cpu_state()}')

# get the data block
db = plc.db_read(DB_NUMBER,START_ADDRESS, SIZE)
# get the start byte of the string to find out its maximum length
string_max_length = db[0]
# copy bytearray of the string, this will be a buffer
string = db[:string_max_length + 2]

new_string = 'Python to TIA'
# write string value to the buffer
set_string(string, 0, new_string, 254)
# now you are ready to write data to the plc
plc.db_write(DB_NUMBER, START_ADDRESS, string)

db = plc.db_read(DB_NUMBER,START_ADDRESS, SIZE)

product_name = get_string(db[0:256], 0, 254)

print(f'Product Name: {product_name}')
product_value=int.from_bytes(db[256:258], byteorder='big')
print(f'Product Value: {product_value}')
product_status=bool(db[258])
print(f'Product Status: {product_status}')

```

**Reply by INDRANI125** [reply]
Thank you for your reply. String writing is happening. Can you help me how can I write integer, bool also? and in case of float how it should be written?

**Reply by nikteliy** [reply]
All data types can be written to the PLC in the same way:

1. Convert a python data type to a PLC data type.
2.  Write it to a block.


Helper functions are located in the [util.py](https://github.com/gijzelaerr/python-snap7/blob/master/snap7/util.py).
Examples can be found in the [example directory](https://github.com/gijzelaerr/python-snap7/tree/master/example).
[Tests](https://github.com/gijzelaerr/python-snap7/tree/master/test) can also be useful as examples.

**Reply by INDRANI125** [reply]
Hi.Thank you but for these example python program.Do we have tia projects too?If yes, can you share that link?

**Reply by swamper123** [reply]
Hey @INDRANI125 
better now than never to respond. ^^
It isn't a TIA Portal thing or that we need an example for that.

On TIA/S7 side, you need to declare how big your `String` is. Otherwise it is 255 chars long as standard, which results in those empty $0 symbols (because nothing/garbage is in there).

In TIA/S7, if you declare `product_name` to `String[20]` it would limit that your name is 20 chars long (and would reduce the zeros at least).

What I haven't tried myself atm. is to set a String and just say it has to be smaller than 255 and is the same length than my actual string.

**Reply by swamper123** [reply]
@INDRANI125 I made a fix #368 if this is still relevant. It was an error on our side.

**Reply by SkilledZero** [reply]
> The first two bytes in the 'String' data type contain additional information about the size of the string. You should use `get_string()` and `set_string()` functions to work with 'String' data type. I modified your code a bit:
> 
> ```
> import snap7
> from snap7.util import set_string, get_string
> 
> IP = '192.168.0.1'
> RACK = 0
> SLOT = 1
> DB_NUMBER = 100
> START_ADDRESS = 0
> SIZE = 259
> 
> plc = snap7.client.Client()
> plc.connect(IP, RACK, SLOT)
> 
> plc_info = plc.get_cpu_info()
> print(f'Module Type: {plc_info.ModuleTypeName}')
> print(f'State: {plc.get_cpu_state()}')
> 
> string_to_db = 'Python to TIA'
> # create a buffer and set buffer size equal to your string length plus 2 bytes for additional info 
> buffer = bytearray(len(string_to_db) + 2)
> # write string value to the buffer
> set_string(buffer, 0, string_to_db, 254)
> # now you are ready to write data to the plc
> plc.db_write(DB_NUMBER, START_ADDRESS, buffer)
> 
> db = plc.db_read(DB_NUMBER,START_ADDRESS, SIZE)
> 
> # get the string value from the String data type
> product_name = get_string(db[0:255], 0, 254)
> 
> print(f'Product Name: {product_name}')
> product_value=int.from_bytes(db[255:257], byteorder='big')
> print(f'Product Value: {product_value}')
> product_status=bool(db[257])
> print(f'Product Status: {product_status}')
> ```

---- im new to github. is this code applicable to s7300? what will be the code if adding more product name like product name 1 product name 2 and product name 3?

**Comment by djosef39** [comment]
Please help me, I'm new to this topic:

When I execute the following instructions:

plc_info = plc.get_cpu_info()
print(f'Module Type: {plc_info.ModuleTypeName}')
print(f'State: {plc.get_cpu_state()}')

I get an error:

b'CPU : Item not available'
Traceback (most recent call last):
  File "C:\Users\ymehda\PycharmProjects\pythonProject1\main.py", line 35, in <module>
    plc_info = plc.get_cpu_info()
  File "C:\Users\ymehda\AppData\Local\Programs\Python\Python39\lib\site-packages\snap7\client.py", line 161, in get_cpu_info
    check_error(result, context="client")
  File "C:\Users\ymehda\AppData\Local\Programs\Python\Python39\lib\site-packages\snap7\common.py", line 92, in check_error
    raise Snap7Exception(error)
snap7.exceptions.Snap7Exception: b'CPU : Item not available'

Does anyone have any idea about this problem?

**Reply by lautarodapin** [reply]
Did you enable PUT/GET in the tia portal for the plc?

**Reply by djosef39** [reply]
Hello, thanks for your reply, 
Yes, PUT/GET in the tia portal for the plc is enable. There are some functions that work, and some of the functions do not work and give an error, as shown in the attached 
![1(1)](https://user-images.githubusercontent.com/87219505/125199553-62193b00-e26f-11eb-9151-f0b1424e4686.png)
![2](https://user-images.githubusercontent.com/87219505/125199562-6a717600-e26f-11eb-8732-e36957a2f2f2.png)
![3](https://user-images.githubusercontent.com/87219505/125199570-72c9b100-e26f-11eb-9cb6-f8ec1c6555c9.png)
pictures

**Reply by lautarodapin** [reply]
Some funciona dont work on PLCs 1200/1500, like the ``get_db``
In the snap7 doc doesn't mention that this function won't work for those models.
Do you have another plc to test it? I don't have plcs at the reach of my hand anymore.
Optimized block access is enabled also right?

**Reply by swamper123** [reply]
From the original snap7 doc:
![grafik](https://user-images.githubusercontent.com/24408657/125238988-50c04500-e2e8-11eb-8a33-d4c318315dc2.png)

**Reply by djosef39** [reply]
thank you very much for your reply, Unfortunately, I don't have another PLC to test, Please send me the original snap7 doc, if you can. djosef39@mail.ru. thanks in advance.

**Reply by swamper123** [reply]
@djosef39 http://snap7.sourceforge.net/

**Comment by denissanga** [comment]
Hi all, I have the same problem. In particular I would use the client.full_upload("DB", 100) function with SIEMENS 1200 but the function seems not available

**Reply by swamper123** [reply]
Hey ho,

![grafik](https://user-images.githubusercontent.com/24408657/171797784-ffd54990-99c1-4915-a37b-401f6e3863a1.png)

It seems, that this is not available for 1200 and 1500.

**Reply by denissanga** [reply]
Hi all, just one more questions. It is possible connect to plc remotely to plc with snap7 ? many thanks

**Reply by swamper123** [reply]
If you can establish a connection to the IPv4 network (via VPN as example), sure.
